Adds support for code formatting with AStyle

kennyweiss opened the pull request:
Details:
* AStyle project link -- http://astyle.sourceforge.net
* Adds option 'ENABLE_ASTYLE' to control whether to search for AStyle.
   As with uncrustify, users can specify an explicit 'ASTYLE_EXECUTABLE' if they do not want to use the one in their path.
* Adds a macro 'blt_add_astyle_target' that can be directly invoked
* Support for Astyle formatting has also been added to the 'blt_add_code_checks' macro, where users can supply an 'ASTYLE_CFG_FILE' rather than an 'UNCRUSTIFY_CFG_FILE' to use AStyle.
* Discovered that earlier versions of Astyle and Uncrustify do not support the dry run option ('MODIFY_FILES' = False), and added comments to the appropriate macros.
   Specifically, Uncrustify v0.61 and later and AStyle v2.05 and later support this option, but earlier versions do not.  BLT does not currently check for this.

Update:
* The AStyle and Uncrustify ``check`` targets are now only generated when supported in their corresponding program.
* Added a wrapper script to run AStyle with return codes indicating success or failure.

kennyweiss commented on cmake/SetupCodeChecks.cmake:
```diff
@@ -226,19 +271,20 @@ endmacro(blt_add_cppcheck_target)
 ## MODIFY_FILES, if set to TRUE, modifies the files in place and adds the created target to
 ## the style target.  Otherwise the files are not modified and the created target is added
 ## to the check target.
+## Note: Setting MODIFY_FILES to FALSE is only supported in Uncrustify v0.61 or greater.
```
Should we add a check for this?

white238 replied:
Not sure how you would...

kennyweiss commented on cmake/SetupCodeChecks.cmake:
```diff
@@ -294,3 +340,89 @@ macro(blt_add_uncrustify_target)
+## MODIFY_FILES, if set to TRUE, modifies the files in place and adds the created target to
+## the style target. Otherwise the files are not modified and the created target is added
+## to the check target.  
+## Note: Setting MODIFY_FILES to FALSE is only supported in AStyle v2.05 or greater.
```
Should we add a check for this?

kennyweiss commented on cmake/SetupCodeChecks.cmake:
```diff
@@ -294,3 +340,89 @@ macro(blt_add_uncrustify_target)
+endif()
+
+add_custom_target(${arg_NAME}
+        COMMAND ${ASTYLE_EXECUTABLE} ${arg_PREPEND_FLAGS}
```
Unlike uncrustify, astyle does not return non-zero when there are files that need to be formatted, so, as long as the config file is valid, these targets passes.

When there are files that need to be formatted, ``astyle`` outputs a line, per file that needs to be formatted of the form:
```cmake
Formatted <filename>
```

Should we add some code to run astyle and grep through the output for these lines so that the behavior is similar to that of our ``uncrustify`` targets?

white238 replied:
That is unfortunate. I know users want it to fail if files need to be formatted.

cyrush replied:
If it's not too hard, we should try to parse and error. 
We will have to weigh with how brittle we feel parsing will be

kennyweiss replied:
I added a wrapper script to run AStyle and parse the results. Both astyle_style and astyle_check targets now fail when there are improperly formatted source files.

white238 commented on cmake/SetupCodeChecks.cmake:
```diff
@@ -294,3 +340,89 @@ macro(blt_add_uncrustify_target)
+macro(blt_add_astyle_target)
+
+## parse the arguments to the macro
+set(options)
```
Indent this macro

kennyweiss commented:
@white238 @cyrush -- I've addressed all the concerns that were raised.

I also removed the code check targets from default builds since they were getting triggered when building an entire solution.

jdahm commented:
On this topic... Could we add others to conform (mostly) with the [GNU standard targets](https://www.gnu.org/prep/standards/html_node/Standard-Targets.html)? It would be nice to have TAGS, uninstall, etc.

kennyweiss commented:
Thanks @jdahm -- seems like an interesting suggestion.  There are definitely good ideas there, but it would be good to discuss what is compatible with cmake/blt, since the backend generator is not always ``make``.

Could you create a separate issue for this so we can discuss/track it?
